File: piper/extensions/chrome/src/engine/statements/text_getSubstring.js
import { stringify } from "../util"



function subsequenceFromStartFromEnd(sequence, at1, at2) {
    let start = at1
    let end = sequence.length - 1 - at2 + 1
    return sequence.slice(start, end)
}

function subsequenceFromStartLast(sequence, at1) {
    let start = at1
    let end = sequence.length - 1 + 1
    return sequence.slice(start, end)
}

function subsequenceFromEndFromStart(sequence, at1, at2) {
    let start = sequence.length - 1 - at1
    let end = at2 + 1
    return sequence.slice(start, end)
}

function subsequenceFromEndFromEnd(sequence, at1, at2) {
    let start = sequence.length - 1 - at1
    let end = sequence.length - 1 - at2 + 1
    return sequence.slice(start, end)
}

function subsequenceFromEndLast(sequence, at1) {
    let start = sequence.length - 1 - at1
    let end = sequence.length - 1 + 1
    return sequence.slice(start, end)
}

function subsequenceFirstFromEnd(sequence, at2) {
    let start = 0
    let end = sequence.length - 1 - at2 + 1
    return sequence.slice(start, end)
}


export default async function(statement, onError) {
    try {
        let pointer = statement.pointer
        let value = await this.executeStatement(stringify(pointer.value))
        value = stringify(value)
        let at1 = await this.executeStatement(stringify(pointer.at1))
        at1 = parseInt(stringify(at1))
        let at2 = await this.executeStatement(stringify(pointer.at2))
        at2 = parseInt(stringify(at2))

        let action1 = stringify(pointer.action1)
        let action2 = stringify(pointer.action2)

        if (action1 === "FROM_START" && action2 === "FROM_START") {
            return value.slice(at1-1, at2)
        } else if (action1 === "FROM_START" && action2 === "FROM_END") {
            return subsequenceFromStartFromEnd(value, at1-1, at2-1)
        } else if (action1 === "FROM_START" && action2 === "LAST") {
            return subsequenceFromStartLast(value, at1-1)
        } else if (action1 === "FROM_END" && action2 === "FROM_START") {
            return subsequenceFromEndFromStart(value, at1-1, at2-1)
        } else if (action1 === "FROM_END" && action2 === "FROM_END") {
            return subsequenceFromEndFromEnd(value, at1-1, at2-1)
        } else if (action1 === "FROM_END" && action2 === "LAST") {
            return subsequenceFromEndLast(value, at1-1)
        } else if (action1 === "FIRST" && action2 === "FROM_START") {
            return value.slice(0, at2)
        } else if (action1 === "FIRST" && action2 === "FROM_END") {
            return subsequenceFirstFromEnd(value, at2-1)
        } else {
            return value
        }
    } catch (e) {
        onError(e)
    }

    return null
}

File: piper/android/app/src/main/java/io/gopiper/piper/engine/statements/text_getSubstring.kt
package io.gopiper.piper.engine.statements

import android.util.Log
import io.gopiper.piper.engine.Piper
import org.json.JSONObject
import java.lang.Exception
import kotlin.math.sqrt

fun subsequenceFromStartFromEnd(sequence: String, at1: Int, at2: Int): String {
    val start = at1
    val end = sequence.length - 1 - at2 + 1
    return sequence.slice(start until end)
}

fun subsequenceFromStartLast(sequence: String, at1: Int): String {
    val start = at1
    val end = sequence.length - 1 + 1
    return sequence.slice(start until end)
}

fun subsequenceFromEndFromStart(sequence: String, at1: Int, at2: Int): String {
    val start = sequence.length - 1 - at1
    val end = at2 + 1
    return sequence.slice(start until end)
}

fun subsequenceFromEndFromEnd(sequence: String, at1: Int, at2: Int): String {
    val start = sequence.length - 1 - at1
    val end = sequence.length - 1 - at2 + 1
    return sequence.slice(start until end)
}

fun subsequenceFromEndLast(sequence: String, at1: Int): String {
    val start = sequence.length - 1 - at1
    val end = sequence.length - 1 + 1
    return sequence.slice(start until end)
}

fun subsequenceFirstFromEnd(sequence: String, at2: Int): String {
    val start = 0
    val end = sequence.length - 1 - at2 + 1
    return sequence.slice(start until end)
}


suspend fun Piper.executeTextGetSubstring(statement: JSONObject, onError: (e: Exception) -> Unit): String? {
    try {
        Log.d("hello", "execute_text_getSubstring")
        val pointer = statement.getJSONObject("pointer")
        val value = executeStatement(pointer["value"].toString()).toString()
        val at1 = executeStatement(pointer["at1"].toString()).toString().toInt()
        val at2 = executeStatement(pointer["at2"].toString()).toString().toInt()

        val action1 = pointer["action1"].toString()
        val action2 = pointer["action2"].toString()


        return if (action1 === "FROM_START" && action2 === "FROM_START") {
            value.slice(at1-1 until at2)
        } else if (action1 === "FROM_START" && action2 === "FROM_END") {
            subsequenceFromStartFromEnd(value, at1-1, at2-1)
        } else if (action1 === "FROM_START" && action2 === "LAST") {
            subsequenceFromStartLast(value, at1-1)
        } else if (action1 === "FROM_END" && action2 === "FROM_START") {
            subsequenceFromEndFromStart(value, at1-1, at2-1)
        } else if (action1 === "FROM_END" && action2 === "FROM_END") {
            subsequenceFromEndFromEnd(value, at1-1, at2-1)
        } else if (action1 === "FROM_END" && action2 === "LAST") {
            subsequenceFromEndLast(value, at1-1)
        } else if (action1 === "FIRST" && action2 === "FROM_START") {
            value.slice(0 until at2)
        } else if (action1 === "FIRST" && action2 === "FROM_END") {
            subsequenceFirstFromEnd(value, at2-1)
        } else {
            value
        }
    } catch (e: Exception) {
        onError(e)
    }
    return null
}